src/main/java/io/cdap/plugin/ariba/source/connector/AribaConnector.java

```diff
@@ -0,0 +1,170 @@
+      count++;
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  @Override
+  public ConnectorSpec generateSpec(ConnectorContext connectorContext, ConnectorSpecRequest connectorSpecRequest)
+    throws IOException {
+    AribaServices aribaServices = new AribaServices(config);
+    try {
+      accessToken = aribaServices.getAccessToken();
+    } catch (AribaException e) {
+      LOG.error("Error in generating token", e);
+    }
+    ConnectorSpec.Builder specBuilder = ConnectorSpec.builder();
+    Map<String, String> properties = new HashMap<>();
+    properties.put(io.cdap.plugin.common.ConfigUtil.NAME_USE_CONNECTION, "true");
+    properties.put(ConfigUtil.NAME_CONNECTION, connectorSpecRequest.getConnectionWithMacro());
+    String template = connectorSpecRequest.getPath();
+    if (template != null) {
+      properties.put(AribaPluginConfig.TEMPLATE_NAME, template);
+    }
+    try {
+      Schema schema = aribaServices.buildOutputSchema(accessToken, template);
+      specBuilder.setSchema(schema);
+    } catch (AribaException | InterruptedException e) {
+      LOG.error("Error in generating schema", e);
+    }
+    return specBuilder.addRelatedPlugin(new PluginSpec(AribaBatchSource.NAME, BatchSource.PLUGIN_TYPE,
+                                                       properties)).build();
+  }
+
+  private JsonArray listTemplates(String pageToken, JsonArray jsonElements) {
+    AribaServices aribaServices = new AribaServices(config);
+    try {
+      accessToken = aribaServices.getAccessToken();
+    } catch (AribaException | IOException e) {
+      LOG.error("unable to generate access token", e);
+    }
+    URL viewTemplatesURL = HttpUrl.parse(config.getBaseURL()).
+      newBuilder()
+      .addPathSegments(METADATA_PATH)
+      .addPathSegments(String.format(PATH_SEGMENT, config.getSystemType()))
+      .addQueryParameter(ResourceConstants.PRODUCT, ResourceConstants.ANALYTICS)
+      .addQueryParameter(ResourceConstants.PAGE_TOKEN, pageToken)
+      .addQueryParameter(ResourceConstants.REALM, config.getRealm()).build().url();
+    AribaResponseContainer responseContainer = null;
+    try {
+      responseContainer = aribaServices.fetchAribaResponse(viewTemplatesURL, accessToken);
+    } catch (AribaException | IOException | InterruptedException e) {
+      LOG.error("Unable to fetch response", e);
+    }
+    InputStream responseStream = responseContainer.getResponseBody();
+    BufferedReader reader = new BufferedReader(new InputStreamReader(responseStream, StandardCharsets.UTF_8));
+    String result = reader.lines().collect(Collectors.joining(""));
+    JsonObject jsonObject = gson.fromJson(result, JsonObject.class);
+    JsonElement nextPageToken = jsonObject.get(TOKEN);
+    try {
+      if (!nextPageToken.isJsonNull()) {
+        listTemplates(nextPageToken.getAsString(), jsonElements);
+        jsonElements.addAll(jsonObject.getAsJsonArray(RECORDS));
+      }
+    } catch (NullPointerException e) {
```

Comment:
why are you catching NullPointerException? This looks like a bug

Reply:
fixed the code.